libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
+                  }
+            }
+      }
+}
+      
+// populate the given harmonys vector with the pitches of the harmony notes implied by the figured base on the given chord
+bool FiguredBass::calcFiguredBass(Chord *chord, vector<int> & harmonys) {
+      Q_ASSERT(harmonys.size() == 0);
+      Segment *segment = chord->segment();
+      int octave = 0; // 0, or 12 if we want the harmony 1 octave up
+      FiguredBassItem::Modifier none = FiguredBassItem::Modifier::NONE;
+      if (find_if(segment->annotations().cbegin(),
+                  segment->annotations().cend(),
+                  [chord](Element *e) {
+                        return ( e->type() == Element::Type::FIGURED_BASS
+                                && e->track() == chord->track());
+                  }) == segment->annotations().cend())
+      {
+             return false; // no figured base to annotate
+      }
+      
+      if ( chord->notes().size() == 0 ) {
+            qDebug("no notes on chord");
+            return false; // no notes in the chord, not sure if this is possible?
+      }
+      if ( chord->notes().size() != 1 ) {
+            qDebug("measure #%d, cannot render Figured Base on a multi-note chord", segment->measure()->no());
+            return false;
+      }
+      Note *note = chord->notes()[0];
+
+      auto mod7 = [] (int n) {
+            n = n%7;
+            while ( n < 0)
+                  n += 7;
+            return n;
+      };
+      auto mod12 = [] (int n) {
+            n = n%12;
+            while ( n < 0)
+                  n += 12;
+            return n;
+      };
+      auto get_digit = [](int digit) {
+            // a modifier with no digit always references the 3rd
+            return (digit == FBIDigitNone ) ? 3 : digit;
+      };
+      auto render_harmony = [&] (FiguredBassItem::Modifier prefix, int relpitch, FiguredBassItem::Modifier suffix) {
+            relpitch = get_digit(relpitch);
+            if ( prefix != none && suffix != none) {
+                  qDebug("(1242) measure #%d, cannot render FiguredBassModifier with two modifiers", segment->measure()->no());
+                  return;
+            }
+            FiguredBassItem::Modifier modifier = (prefix == none) ? suffix : prefix;
+            ClefType clef = chord->staff()->clef(segment->tick());
+            
+            // calculate native pitch for line adjusted for transposing instrument.
+            // in doing so we have to SUBTRACT(relpitch-1) because lines are oriented from top to bottom, and
+            // pitches are oriented from bottom to top.  The 1 is because chord harmonies are noted as 1,3,5 (for a normal triad)
+            // rather than (0,2,4).
+            int harmony_line = note->line() - relpitch + 1;
+            int native_pitch = line2pitch(harmony_line, clef, Key::C) + note->ppitch() - note->epitch();
+
+            AccidentalVal acciv = segment->measure()->findAccidental(segment, chord->staff()->idx(), harmony_line);
```

Comment:
The reason I calculate in terms of line is so I can add integers to calculate the intervals.
This is not possible with pitches, because of diatonic scale issues.  There are functions which calculate offsets within diatonic scales, but they explicitly ignore accidentals which occur earlier in the measure.  The findAccidental function understands both key signature and also accidentals, explicitly obeying musescore rules for which accidentals affect future notes and which do not, such as ties, and note occurring in other staffs.


> On 24 May 2015, at 05:00, Maurizio M. Gavioli <notifications@github.com> wrote:
> 
> In libmscore/figuredbass.cpp <https://github.com/musescore/MuseScore/pull/2015#discussion_r30949763>:
> 
> > +            relpitch = get_digit(relpitch);
> > +            if ( prefix != none && suffix != none) {
> > +                  qDebug("(1242) measure #%d, cannot render FiguredBassModifier with two modifiers", segment->measure()->no());
> > +                  return;
> > +            }
> > +            FiguredBassItem::Modifier modifier = (prefix == none) ? suffix : prefix;
> > +            ClefType clef = chord->staff()->clef(segment->tick());
> > +            
> > +            // calculate native pitch for line adjusted for transposing instrument.
> > +            // in doing so we have to SUBTRACT(relpitch-1) because lines are oriented from top to bottom, and
> > +            // pitches are oriented from bottom to top.  The 1 is because chord harmonies are noted as 1,3,5 (for a normal triad)
> > +            // rather than (0,2,4).
> > +            int harmony_line = note->line() - relpitch + 1;
> > +            int native_pitch = line2pitch(harmony_line, clef, Key::C) + note->ppitch() - note->epitch();
> > +
> > +            AccidentalVal acciv = segment->measure()->findAccidental(segment, chord->staff()->idx(), harmony_line);
> Using note pitch rather than line would take any previous accidental into account.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/2015/files#r30949763>.
>